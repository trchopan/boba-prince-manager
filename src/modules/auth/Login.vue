<template>
  <router-view />
</template>

<script>
import { mapGetters } from "vuex";

const LANGUAGES_MAP = {};

export default {
  name: "Login",
  data() {
    return {
      email: "",
      password: ""
    };
  },
  computed: {
    ...mapGetters({
      error: "error",
      authUser: "auth/authUser",
      loading: "auth/loading",
      language: "layout/language"
    }),
    $t() {
      return this.$translate(LANGUAGES_MAP, this.language.value);
    },
    errorMessage() {
      const lastError = this.error[this.error.length - 1];
      if (!lastError || !lastError.code) {
        return "";
      }
      switch (lastError.code) {
        case "auth/user-not-found":
          return this.$t.userNotFound;
        case "auth/invalid-email":
          return this.$t.invalidEmail;
        default:
          return "";
      }
    }
  }
};
</script>
