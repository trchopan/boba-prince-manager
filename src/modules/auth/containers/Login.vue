<template>
  <v-layout align-center justify-center>
    <v-flex xs12 sm8 md4>
      <v-card class="elevation-12">
        <v-toolbar dark color="primary">
          <v-toolbar-title>{{ $t.login }}</v-toolbar-title>
        </v-toolbar>
        <v-progress-linear
          :active="loading"
          indeterminate
          color="orange"
          height="3"
        />
        <v-form>
          <v-card-text>
            <v-text-field
              id="email"
              prepend-icon="person"
              name="email"
              label="Email"
              type="text"
              :error-messages="error ? error.message : ''"
              v-model="email"
            />
            <v-text-field
              id="password"
              prepend-icon="lock"
              name="password"
              :label="$t.password"
              type="password"
              v-model="password"
            />
          </v-card-text>
          <v-card-actions>
            <v-spacer/>
            <v-btn
              id="submit"
              color="primary"
              type="submit"
              :disabled="loading"
              @click.prevent="onSubmit()"
            >{{ $t.login }}</v-btn>
          </v-card-actions>
        </v-form>
      </v-card>
    </v-flex>
  </v-layout>
</template>

<script>
import { mapGetters } from "vuex";
import { translate } from "@/share/helpers";

const languagesMap = {
  login: { vi: "Đăng nhập", en: "Log in" },
  password: { vi: "Mật khẩu", en: "Password" }
};

export default {
  name: "Login",
  data() {
    return {
      email: "",
      password: ""
    };
  },
  computed: {
    ...mapGetters({
      error: "auth/error",
      authUser: "auth/authUser",
      loading: "auth/loading",
      language: "layout/language"
    }),
    $t() {
      return translate(languagesMap, this.language.code);
    }
  },
  methods: {
    async onSubmit() {
      const success = await this.$store.dispatch(
        "auth/loginWithEmailPassword",
        {
          email: this.email,
          password: this.password
        }
      );
      if (success) this.$router.replace("/");
    }
  }
};
</script>

<style lang="scss" scoped>
.v-progress-linear {
  margin: 0;
}
</style>
