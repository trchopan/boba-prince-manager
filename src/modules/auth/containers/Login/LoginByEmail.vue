<template>
  <v-layout align-center justify-center>
    <v-flex xs12 sm6 md4>
      <v-card class="elevation-12">
        <v-form>
          <v-card-text>
            <v-text-field
              autofocus
              id="email"
              prepend-icon="person"
              name="email"
              label="Email"
              type="text"
              :error-messages="errorMessage"
              v-model="email"
            />
            <v-text-field
              id="password"
              prepend-icon="lock"
              name="password"
              :label="$t.password"
              type="password"
              v-model="password"
            />
          </v-card-text>
          <v-card-actions>
            <v-spacer/>
            <v-btn
              outline
              to="/signup"
              color="secondary"
              type="button"
            >{{ $t.signUp }}</v-btn>
            <v-btn
              id="submit"
              color="primary"
              type="submit"
              :disabled="loading.login"
              @click.prevent="onSubmit()"
            >{{ $t.login }}</v-btn>
          </v-card-actions>
        </v-form>
      </v-card>
    </v-flex>
  </v-layout>
</template>

<script>
import { mapGetters } from "vuex";

const LANGUAGES_MAP = {
  signUp: { vi: "Đăng ký", en: "Sign up" },
  login: { vi: "Đăng nhập", en: "Log in" },
  password: { vi: "Mật khẩu", en: "Password" },
  userNotFound: {
    vi: "Không tìm thấy tài khoản. Bạn vui lòng kiểm tra lại thông tin",
    en: "Account not found. Please check your information again."
  },
  invalidEmail: { vi: "Email không hợp lệ", en: "Invalid email" }
};

export default {
  name: "Login",
  data() {
    return {
      email: "",
      password: ""
    };
  },
  computed: {
    ...mapGetters({
      error: "error",
      authUser: "auth/authUser",
      loading: "auth/loading",
      language: "layout/language"
    }),
    $t() {
      return this.$translate(LANGUAGES_MAP, this.language.value);
    },
    errorMessage() {
      const lastError = this.error[this.error.length - 1];
      if (!lastError || !lastError.code) {
        return "";
      }
      switch (lastError.code) {
        case "auth/user-not-found":
          return this.$t.userNotFound;
        case "auth/invalid-email":
          return this.$t.invalidEmail;
        default:
          return "";
      }
    }
  },
  methods: {
    async onSubmit() {
      const success = await this.$store.dispatch(
        "auth/loginWithEmailPassword",
        {
          email: this.email,
          password: this.password
        }
      );
      if (success) this.$router.replace("/");
    }
  }
};
</script>
