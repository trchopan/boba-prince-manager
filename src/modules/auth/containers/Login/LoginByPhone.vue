<template>
  <v-layout
    align-center
    justify-center
  >
    <v-flex
      xs12
      sm6
      md4
    >
      <transition
        name="fade"
        mode="out-in"
      >
        <v-card
          v-if="!confirmationResult"
          key="phone-number"
          class="elevation-12"
        >
          <v-form>
            <v-card-text>
              <p>{{ $t.descriptionPhoneNumber }}</p>
              <v-select
                outline
                :value="selectedCountry"
                :items="phoneCountries"
                :label="$t.selectCountry"
              ></v-select>
              <v-text-field
                autofocus
                outline
                required
                prepend-inner-icon="phone"
                name="phone"
                mask="(###) ### - #####"
                :label="$t.phoneNumber"
                type="tel"
                v-model="phoneNumber"
              />
              <p v-show="errorMessage">{{ errorMessage }}</p>
            </v-card-text>
            <v-card-actions>
              <v-spacer/>
              <v-btn
                id="submit"
                color="primary"
                type="submit"
                :disabled="loading.login"
                @click.prevent="onNext()"
              >{{ $t.next }}</v-btn>
            </v-card-actions>
          </v-form>
        </v-card>
        <v-card
          v-if="confirmationResult"
          key="phone-confirmation-result"
        >
          <v-form>
            <v-card-text>
              <p>{{ $format($t.descriptionConfirmationResult, [phoneNumber]) }}</p>
              <v-text-field
                autofocus
                outline
                required
                prepend-inner-icon="lock"
                name="lock"
                mask="######"
                :label="$t.confirmationCode"
                type="phone"
                v-model="confirmationCode"
              />
              <p v-show="errorMessage">{{ errorMessage }}</p>
            </v-card-text>
            <v-card-actions>
              <v-spacer/>
              <v-btn
                id="confirm-code"
                color="primary"
                type="submit"
                :disabled="loading.login"
                @click.prevent="onConfirmCode()"
              >{{ $t.verify }}</v-btn>
            </v-card-actions>
          </v-form>
        </v-card>
      </transition>
    </v-flex>
    <div id="recaptcha-container"></div>
  </v-layout>
</template>

<script>
import { mapGetters } from "vuex";
import { fireAuth, ReCaptchaVerifier } from "@/firebase.js";
import { phoneCountries } from "@/modules/auth/auth.models.js";

const LANGUAGES_MAP = {
  descriptionPhoneNumber: {
    vi:
      "Bạn sẽ nhận tin nhắn SMS vào số điện thoại cung cấp ở trên. Bạn vui lòng điền số này mã xác thực ở cửa sổ tiếp theo.",
    en:
      "You will receive a SMS text message to the phone number provied. Please enter this code in to the confirmation in the next state."
  },
  descriptionConfirmationResult: {
    vi:
      "Tin nhắn mã xác nhận đã được gởi vào số điện thoại {0}. Bạn vui lòng điền mã bên dưới",
    en:
      "A confirmation code has been sent to your number {0}. Please enter the code below."
  },
  selectCountry: { vi: "Chọn mã vùng", en: "Select country" },
  phoneNumber: { vi: "Số điện thoại", en: "Phone number" },
  confirmationCode: { vi: "Mã xác nhận", en: "Confirmation code" },
  next: { vi: "Tiếp tục", en: "Next" },
  verify: { vi: "Xác nhận", en: "Verify" },
  invalidPhone: { vi: "Số điện thoại không hợp lệ", en: "Invalid Phone Number" }
};

export default {
  name: "LoginByPhone",
  data() {
    return {
      recaptcha: null,
      confirmationResult: null,
      phoneCountries: Object.values(phoneCountries),
      selectedCountry: phoneCountries.vn,
      confirmationCode: "",
      phoneNumber: ""
    };
  },
  computed: {
    ...mapGetters({
      error: "error",
      authUser: "auth/authUser",
      loading: "auth/loading",
      language: "language"
    }),
    $t() {
      return this.$translate(LANGUAGES_MAP, this.language.value);
    },
    errorMessage() {
      const lastError = this.error[this.error.length - 1];
      if (!lastError || !lastError.code) {
        return "";
      }
      switch (lastError.code) {
        default:
          return this.$t.unknownError;
      }
    }
  },
  methods: {
    async onNext() {
      const phoneNumber =
        this.selectedCountry.value + this.phoneNumber.replace(/^0/, "");
      try {
        this.confirmationResult = await fireAuth.signInWithPhoneNumber(
          phoneNumber,
          this.recaptcha
        );
        console.log(this.confirmationResult);
      } catch (error) {
        console.log("ERROR", error);
      }
    },
    async onConfirmCode() {
      if (!this.confirmationCode) {
        console.log("woot no code?", this.confirmationCode);
        return;
      }
      const user = await this.confirmationResult.confirm(this.confirmationCode);
      if (!user) {
        console.log("naniii no user");
      }
    }
  },
  mounted() {
    this.recaptcha = new ReCaptchaVerifier("recaptcha-container", {
      size: "invisible",
      callback: res => {
        console.log("woot", res);
      }
    });
  }
};
</script>

<style lang="scss" scoped>
#recaptcha-container {
  display: none;
}
</style>
