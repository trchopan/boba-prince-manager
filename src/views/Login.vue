<template>
  <v-layout align-center justify-center>
    <v-flex xs12 sm8 md4>
      <v-card class="elevation-12">
        <v-toolbar dark color="primary">
          <v-toolbar-title>Login form</v-toolbar-title>
        </v-toolbar>
        <v-form>
          <v-card-text>
            <v-text-field
              id="email"
              prepend-icon="person"
              name="email"
              label="Email"
              type="text"
              :error-messages="error ? error.message : ''"
              v-model="email" />
            <v-text-field
              id="password"
              prepend-icon="lock"
              name="password"
              label="Password"
              type="password"
              v-model="password" />
          </v-card-text>
          <v-card-actions>
            <v-spacer />
            <v-btn
              id="submit"
              color="primary"
              type="submit"
              :disabled="loading"
              @click.prevent="onSubmit()">
              Login
            </v-btn>
          </v-card-actions>
        </v-form>
      </v-card>
    </v-flex>
  </v-layout>
</template>

<script>
import { mapGetters } from "vuex";

export default {
  name: "Login",
  data() {
    return {
      email: "",
      password: ""
    };
  },
  computed: mapGetters({
    error: "auth/error",
    authUser: "auth/authUser",
    loading: "auth/loading"
  }),
  methods: {
    async onSubmit() {
      const success = await this.$store.dispatch(
        "auth/loginWithEmailPassword",
        {
          email: this.email,
          password: this.password
        }
      );
      if (success) this.$router.replace("/");
    }
  }
};
</script>
