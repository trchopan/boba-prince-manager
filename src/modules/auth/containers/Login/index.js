import Login from "./Login.vue";
import LoginByPhone from "./LoginByPhone.vue";

export { Login, LoginByPhone };
