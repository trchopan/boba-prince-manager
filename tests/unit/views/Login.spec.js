import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Login from "@/views/Login.vue";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("views/Login.vue", () => {
  let state;
  let getters;
  let actions;
  let mutations;
  let store;
  let routes;
  let router;
  let fakeDepsSuccess;

  beforeEach(() => {
    // Setup the store module "auth"
    state = {
      currentUser: null
    };
    getters = {
      "auth/isLoggedIn": jest.fn(
        () => (state.currentUser && state.currentUser.uid ? true : false)
      )
    };
    actions = {
      "auth/loginWithEmailPassword": jest.fn()
    };
    mutations = {
      "auth/loggedIn": state => (state.currentUser = { uid: "1234" })
    };
    store = new Vuex.Store({
      state,
      getters,
      actions,
      mutations
    });
  });

  it("connects the form with user inputs when submit", () => {
    const wrapper = shallowMount(Login, { store, localVue });
    const form = wrapper.find("form");
    const email = wrapper.find('form > input[type="text"]');
    const password = wrapper.find('form > input[type="password"]');
    email.element.value = "email";
    email.trigger("input");
    password.element.value = "password";
    password.trigger("input");
    expect(wrapper.vm.$data.email).toBe("email");
    expect(wrapper.vm.$data.password).toBe("password");
    form.trigger("submit");
    expect(actions["auth/loginWithEmailPassword"].mock.calls[0][1]).toEqual({
      email: email.element.value,
      password: password.element.value
    });
  });

  it("redirects to home when logged in", () => {
    let mockPush = jest.fn();
    const wrapper = shallowMount(Login, {
      mocks: { $router: { push: mockPush } },
      store,
      localVue
    });
    store.commit("auth/loggedIn");
    return wrapper.vm.$nextTick().then(next => {
      expect(getters["auth/isLoggedIn"].mock.calls.length).toBe(2);
      expect(mockPush.mock.calls).toEqual([["/home"]]);
    });
  });

  it.skip("matchs snapshot", () => {
    const wrapper = shallowMount(Login, {});
    expect(wrapper.element).toMatchSnapshot();
  });
});
