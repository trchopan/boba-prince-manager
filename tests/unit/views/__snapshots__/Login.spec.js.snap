// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/Login.vue matchs snapshot 1`] = `
<v-layout-stub align-center="" justify-center="">
  <v-flex-stub xs12="" sm8="" md4="">
    <v-card-stub class="elevation-12">
      <v-toolbar-stub dark="" color="primary">
        <v-toolbar-title-stub>Login form</v-toolbar-title-stub>
      </v-toolbar-stub>
      <v-progress-linear-stub indeterminate="" dark="" color="orange" height="3"></v-progress-linear-stub>
      <v-form-stub>
        <v-card-text-stub>
          <v-text-field-stub id="email" prepend-icon="person" name="email" label="Email" type="text" error-messages=""></v-text-field-stub>
          <v-text-field-stub id="password" prepend-icon="lock" name="password" label="Password" type="password"></v-text-field-stub>
        </v-card-text-stub>
        <v-card-actions-stub>
          <v-spacer-stub></v-spacer-stub>
          <v-btn-stub id="submit" color="primary" type="submit">
            Login
          </v-btn-stub>
        </v-card-actions-stub>
      </v-form-stub>
    </v-card-stub>
  </v-flex-stub>
</v-layout-stub>
`;
